### DISCUSSION TITLE
Calculating time taken in uefi application.

### DISCUSSION THREAD

**Discussion by ayushdas2000** [Kostr/UEFI-Lessons#5] [General]
Hi Kostr. Thanks for the lessons first.I want to calculate how much time does my application take to run. I am trying to use ticks to calculate time elapsed. Basically this using the PerformanceLib and TimerLib :
Frequency = GetPerformanceCounterProperties(NULL, NULL);
StartTick = GetPerformanceCounter();
EndTick = GetPerformanceCounter();
ElapsedTicks = EndTick > StartTick ? EndTick - StartTick : StartTick - EndTick;
TimeInSeconds = (double)ElapsedTicks / Frequency;
I am getting 0 seconds everytime. Any leads for this ? Thanks in advance.

**Comment by ayushdas2000** [comment]
Tried a lot of things like printing in nanosecs and as string but always getting 0. Any other approach would also help. Thanks

**Reply by Kostr** [reply]
Which library instance do you use for the `TimerLib` library class?

**Reply by ayushdas2000** [reply]
Sorry . not using the timerlib. Only The performance lib for start and end tick. Do i need to intialize something first for this?

**Reply by Kostr** [reply]
Which library instance do you use for the `PerformanceLib` library class then? You should have something like `PerformanceLib|<...>` in your DSC

**Reply by ayushdas2000** [reply]
PerformanceLib|MdePkg/Library/BasePerformanceLibNull/BasePerformanceLibNull.inf

**Reply by Kostr** [reply]
This is an instance of Performance Library with empty functions https://github.com/tianocore/edk2/blob/master/MdePkg/Library/BasePerformanceLibNull/BasePerformanceLibNull.inf

You should replace it with something like `MdeModulePkg/Library/DxePerformanceLib/DxePerformanceLib.inf`

**Reply by Kostr** [reply]
You can find possible library instances with something like 

```grep "PerformanceLib" -r ./ --include="*.inf" | grep LIBRARY_CLASS```

**Reply by ayushdas2000** [reply]
error 1001: Module type [PEIM] is not supported by library instance [/u/aydas/last/edk2/MdeModulePkg/Library/DxePerformanceLib/DxePerformanceLib.inf]
        consumed by library instance [/u/aydas/last/edk2/MdeModulePkg/Core/DxeIplPeim/DxeIpl.inf] which is consumed by module [/u/aydas/last/edk2/MdeModulePkg/Core/DxeIplPeim/DxeIpl.inf]
        
        Could you look at this error please. Thanks

**Reply by Kostr** [reply]
Is the module where you try to use these library functions a PEI module?

**Reply by ayushdas2000** [reply]
No module type is UEFI application

**Reply by ayushdas2000** [reply]
Figured it out finally . Thanks

**Reply by Kostr** [reply]
So what was the problem in your case?
